File: c_solutions_11-20/Euler_12.c
#include "euler_util.h"


int main(int argc, char *argv[])
{
    float start = timeit();

    int base = 1001;
    int divisors = 2;

    long long tri_num = 0;
    long double root;

    for (int i=1; i <= base; i++) 
        tri_num += i;

    while (divisors <= 500) {
        root = sqrt(tri_num);
        for (int i=2; i <= root + 1; i++) {
            if (tri_num % i == 0) 
                divisors += 2;    
        }

        if (divisors <= 500) {
            tri_num += (++base);
            divisors = 2;
        }
    } 
  
    float stop = timeit();

    printf ("Answer: %lld\n", tri_num);
    printf ("Time: %f\n", stop - start);

    return 0;
}

File: haskell_11-20/Euler_12.hs
-- First triangle number with +500 divisors

module Main where
import Text.Printf
import System.Clock


isDivisible :: Int -> Int -> Bool
isDivisble _ 0  = False
isDivisible x y = x `mod` y == 0

range :: Int -> Int
range n = (ceiling $ sqrt $ fromIntegral n) + 1

-- keep track of the "range n" passing in an extra [Int] parameter
divisors :: Int -> Int
divisors n = sum [ 1 | x <- [2..range n], isDivisible n x] * 2

triangle :: Int -> Int -> Int -> Int
triangle n c lim
    | divs > lim = n
    | otherwise  = triangle (n + c) (c + 1) lim
  where
    divs = divisors n

getTotalTime :: TimeSpec -> Float
getTotalTime t = fromIntegral (sec t) + (fromIntegral(nsec t) / 10^9)

main = 
  do
    start <- getTime Monotonic
    printf "Answer: %d\n" (triangle 3 3 500)
    stop <- getTime Monotonic
    printf "Time:   %.4f\n" (getTotalTime(stop) - getTotalTime(start))
